FIX DOC MNT Big revamp of cross-decomposition module

NicolasHug opened the pull request:
Closes https://github.com/scikit-learn/scikit-learn/issues/4122
Closes #8392
Probably Fixes https://github.com/scikit-learn/scikit-learn/issues/4469, though I can't reproduce.
Fixes https://github.com/scikit-learn/scikit-learn/issues/11645
Closes https://github.com/scikit-learn/scikit-learn/issues/13521
Closes https://github.com/scikit-learn/scikit-learn/issues/16177

This PR is a rework of the cross-decomposition module which had been left for dead for years. Mainly, docs and tests were added, and code was simplified. This hopefully comes with a few bug fixes.

This is a big PR, but it's a lot of docs. You might want to ignore the diff and just review the files from scratch, given the amount of changes. The good news is that there *are* docs now, so you're off to a much better start than I was.

Other stuff:

- use 1d shapes for vectors instead of 2d shapes. Makes outer product more obvious.


- for PLSSVD, CCA, PLSCanonical `n_components` now raises a FutureWarning if it's not in `[1, min(n_samples, n_features, n_targets)]`. This is only for backward compat, and an error will be raised in 2 versions: `n_components` cannot be greater than the rank of the cross-covariance matrix = X.T.dot(Y) which is bounded as above. For PLSRegression, the rank is bounded by n_features. See comments in code.

- PLSSVD, CCA and PLSCanonical:
	- deprecated x_scores and y_scores attributes since they're useless: they're just the transformed matrices of `X` and `Y`. For PLSRegression, y_scores are different, so I didn't deprecate (But I doubt these are useful anyway)

NicolasHug committed:
Simplified array shapes

NicolasHug committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into plsmaybe

NicolasHug committed:
improved and cleaned tests

NicolasHug committed:
Cleaned up PLSSVD

NicolasHug committed:
Some cleanings

NicolasHug committed:
Some more cleaning

NicolasHug committed:
Correct bounds for n_components

NicolasHug committed:
should fix warnings

NicolasHug committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into la_PLS_qui_porte_si_bien_son_nom

NicolasHug committed:
removed norm_y_weight

NicolasHug committed:
WIP

NicolasHug committed:
deprecated scores

NicolasHug committed:
WIP

NicolasHug committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into la_PLS_qui_porte_si_bien_son_nom

NicolasHug committed:
some docs

NicolasHug committed:
some more

NicolasHug committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into la_PLS_qui_porte_si_bien_son_nom

NicolasHug committed:
fixed docstring tests

NicolasHug commented:
CC @thomasjpfan 

@amueller you have time for reviews now, right? Welcome back!! :p

TomDLT commented on sklearn/cross_decomposition/_pls.py:
```diff
@@ -135,128 +116,35 @@ def _center_scale_xy(X, Y, scale=True):
     return X, Y, x_mean, y_mean, x_std, y_std
 
 
+def _svd_flip_1d(u, v):
```
You should add a test to make sure that it is the same as `svd_flip`.

TomDLT reviewed (approved):
Nice cleaning !

Do we need a test for #11645 ?

NicolasHug commented on sklearn/cross_decomposition/_pls.py:
```diff
@@ -883,62 +844,87 @@ def fit(self, X, Y):
     def transform(self, X, Y=None):
         """
-        Apply the dimension reduction learned on the train data.
+        Apply the dimensionality reduction learned on the training data.
```
no but this just says "learned on the training data"

TomDLT replied:
Oh I see, I read it "Apply (the dimensionality reduction learned) on the training data.".
I would remove "on the training data" then.

NicolasHug committed:
Apply suggestions from code review

Co-authored-by: Tom Dupré la Tour <tom.dupre-la-tour@m4x.org>

NicolasHug committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into la_PLS_qui_porte_si_bien_son_nom

NicolasHug committed:
Merge branch 'la_PLS_qui_porte_si_bien_son_nom' of github.com:NicolasHug/scikit-learn into la_PLS_qui_porte_si_bien_son_nom

NicolasHug commented on doc/modules/cross_decomposition.rst:
```diff
@@ -23,19 +20,167 @@ Cross decomposition algorithms find the fundamental relations between two
+function of the tranformed samples :math:`\Xi`, by computing :math:`\alpha
+\in \mathcal{R}` such that :math:`\Omega = \alpha \Xi`.
+
+Then, we have :math:`Y = \Omega \Gamma^T = \alpha \Xi \Gamma^T`, and since
```
yes thanks!

NicolasHug committed:
Addressed comments

NicolasHug committed:
added test for svd_flip_1d

NicolasHug committed:
some more

NicolasHug commented:
Thanks @TomDLT ! I think I addressed all the comments

> Do we need a test for #11645 ?

I don't think so: I literally removed the use of `svds`. I'm not quite sure why it was previously used honestly. Given the state of the code (old and unmaintained for years), I went for a simple solution: if I don't understand the reason and if it doesn't break the tests, then I'm removing.

TomDLT commented:
Fair, thanks for this great cleaning !

amueller commented:
Maybe @agramfort or @bthirion can have a look? (if I remember correctly they know something about this).

NicolasHug commented:
@duchesnay since you're the original author, your input would be very welcome too!

NicolasHug committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into la_PLS_qui_porte_si_bien_son_nom

NicolasHug committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into la_PLS_qui_porte_si_bien_son_nom

NicolasHug committed:
added to ignore docstring checks

NicolasHug committed:
typo

agramfort commented on doc/modules/cross_decomposition.rst:
```diff
@@ -23,19 +20,170 @@ Cross decomposition algorithms find the fundamental relations between two
+projects both `X` and `Y` into a lower-dimensional subspace such that the
+covariance between `tranformed(X)` and `transformed(Y)` is maximal.
+
+PLS draws similarities with Principal Component Regression, where the samples
```
```suggestion
PLS draws similarities with Principal Component Regression (PCR), where the samples
```

agramfort replied:
do we say somewhere what PCR does (pipeline PCA -> Regressor) or do we provide a ref?

agramfort commented on doc/modules/cross_decomposition.rst:
```diff
@@ -23,19 +20,170 @@ Cross decomposition algorithms find the fundamental relations between two
-Classes included in this module are :class:`PLSRegression`
+Apart from CCA, the PLS estimators are particularly suited when the matrix of
+predictors has more variables than observations, and when there is
+multicollinearity among the features. By contrast, standard linear regression
```
The fact that it fixes the multicollinearity issue while linear regression would fail unless regularization seems an authoritative argument here. Do we have a ref or example that demo this?

NicolasHug replied:
When features are collinear, the covariance matrix is ill-defined and thus non-invertible. 

I don't have a ref at hand, but I'm assuming this is common knowledge (could be wrong?)

agramfort replied:
are you referring to this https://github.com/scikit-learn/scikit-learn/pull/17095/files#diff-df97917f68917d3a110df30940d771dfR176 ? if it's a statement of PLS vs CCA rather than PLS vs linear regression. Maybe I am nitpicking here.

NicolasHug replied:
It's mostly a statement about PLS vs non-regularized LR: LR is unstable when there is collinearity among features, PLS is not. However, CCA is unstable too (as described in the link you mentioned).

agramfort commented on doc/modules/cross_decomposition.rst:
```diff
@@ -23,19 +20,170 @@ Cross decomposition algorithms find the fundamental relations between two
+
+Given two centered matrices :math:`X \in \mathbb{R}^{n \times d}` and
+:math:`Y \in \mathbb{R}^{n \times t}`, and a number of components :math:`K`,
+:class:`PLSCanonical` procedes as follows:
```
```suggestion
:class:`PLSCanonical` proceeds as follows:
```

NicolasHug committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into la_PLS_qui_porte_si_bien_son_nom

NicolasHug committed:
Addressed comments from Alex

agramfort reviewed (approved):
thx @NicolasHug for clarifying

NicolasHug commented:
Thanks @TomDLT and @agramfort for the reviews!

Let me push a what'snew entry, and I'll merge when green

NicolasHug committed:
Added whatsnew + minor cosmetics

NicolasHug committed:
typo

NicolasHug committed:
fixed docs?

NicolasHug commented:
Docs look good, merging.
Thanks for the reviews!

hendriklohse commented:
@NicolasHug I have a question about the documentation. What exactly does the attribute x_loadings_ represent? The correlation, or the coefficients of the linear combinations used to transform X? (See also https://stackoverflow.com/questions/78725061/what-does-the-attribute-x-loadings-represent?noredirect=1#comment138800857_78725061)
